Add velocity control mode

dogoepp opened the pull request:
Here a new control mode is supported : sending commands to servos in velocity command mode (wheel in Robotis' jargon).

The mode is detected in the servos but we require that the user declares the desired control mode with a ros parameter, to be sure to detect wrong assumption on the current control mode of a servo.

dogoepp committed:
Add velocity joints

dogoepp committed:
move some pieces of code to their own functions

dogoepp committed:
added comments

dogoepp committed:
reorder parameters for more coherence

dogoepp committed:
little fix

costashatz commented:
> Here a new control mode is supported : sending commands to servos in velocity command mode (wheel in Robotis' jargon).

Thanks for this nice PR..

> the user declares the desired control mode with a ros parameter

Is this really necessary? Or you did it for safety?

dogoepp commented:
For the sole sake of safety I added this limitation to user-friendliness. I still wonder if it is a good idea.

dogoepp committed:
Stricter checks on command interface mode

dogoepp committed:
Re-factor parameter structure.

It allows for easier reading and to set a default interface mode.

dogoepp committed:
Update to current status

costashatz commented:
This overall seems nice! Only one question/remark: `ROS_ERROR_STREAM` and `ROS_FATAL_STREAM` do not exit the program right? I think we should terminate the program, especially for the `ROS_FATAL_STREAM` cases..

costashatz reviewed (commented):
This overall seems nice! Only one question/remark: ROS_ERROR_STREAM and ROS_FATAL_STREAM do not exit the program right? I think we should terminate the program, especially for the ROS_FATAL_STREAM cases..

dogoepp commented:
You're right for the fatal that should exit.

For the errors related to command interface type (position/velocity), the joints concerned are disabled. Is that not enough ? Do we need to give such a strong signal as to crash the whole program ?

dogoepp committed:
fatal error makes the program stop

costashatz commented on src/dynamixel_control_hw.cpp:
```diff
@@ -62,55 +71,94 @@ int main(int argc, char** argv)
+    }
+    catch (XmlRpc::XmlRpcException& e) {
+        ROS_FATAL_STREAM("Exception raised when parsing the configuration: " << e.getMessage());
+        return 1;
```
Most probably naive question, but just making sure. This should be enough to not allow incorrect configurations right? I.e., nothing is needed in the `dynamixel_hardware_interface.hpp` file. Just making sure..

dogoepp replied:
What is an incorrect configuration for you ?

The above code should only accept well-structures configuration. However, it does not prevent the addition of unused parameters (which are then ignored).
If a parameter is expected to be an integer but is of an other type, an exception is raised and causes the program to stop. Saddly, the exception message is not informative. I don't know how to improve this, except by splitting this try/catch in more, smaller ones.

costashatz replied:
> The above code should only accept well-structures configuration.

That's what I correct configuration is for me..

> If a parameter is expected to be an integer but is of an other type, an exception is raised and causes the program to stop.

And that's what an incorrect configuration is for me. :smile: 

> However, it does not prevent the addition of unused parameters (which are then ignored).

This is ok.

> Saddly, the exception message is not informative. I don't know how to improve this, except by splitting this try/catch in more, smaller ones.

I think we should not bother too much for this.. It is not of that big importance..

Great! This is ok for me then!

costashatz commented on include/dynamixel_control_hw/dynamixel_hardware_interface.hpp:
```diff
@@ -212,47 +180,61 @@ namespace dynamixel {
+
+                    // enable torque output on the servo and set its configuration
+                    // including max speed
+                    _enable_and_configure_servo(_servos[i], hardware_mode);
```
If I am not mistaken, when the `hardware_mode` is `OperatingModel::unknown` we should not enable the servo but from what I see we still `set_torque_enable(1)` in this function (`_enable_and_configure_servo`) no matter the mode. Of course, I might be missing something..

dogoepp replied:
You're right. Fixed

costashatz reviewed (commented):
I like how this code is becoming. Good job @dogoepp! Some very minor comments...

dogoepp committed:
Error message hopefully more informative.

dogoepp committed:
completely disable mis-configured actuators.
In this case, we are talking about actuators which operating mode
(position or velocity) is not declared, improperly declared or does not
match the hardware setting.

costashatz committed:
Better gitignore

costashatz committed:
Remove also servos that were not found in params

costashatz reviewed (approved):
@dogoepp this should be good to merge! Great job! Can you only check my last commit 
864bf9f ?

dogoepp commented:
Looks good. Thanks !

**Merging**
